File: backend/app/supletorios/business_days.py
from datetime import date, timedelta

COLOMBIAN_HOLIDAYS: set[date] = {
    # 2025
    date(2025, 1, 1),    # Año Nuevo
    date(2025, 1, 6),    # Día de los Reyes Magos
    date(2025, 3, 24),   # Lunes Santo
    date(2025, 4, 17),   # Jueves Santo
    date(2025, 4, 18),   # Viernes Santo
    date(2025, 5, 1),    # Día del Trabajo
    date(2025, 5, 26),   # Ascensión del Señor
    date(2025, 6, 16),   # Corpus Christi
    date(2025, 6, 23),   # Sagrado Corazón
    date(2025, 6, 29),   # San Pedro y San Pablo
    date(2025, 7, 20),   # Día de la Independencia
    date(2025, 8, 7),    # Batalla de Boyacá
    date(2025, 8, 18),   # Asunción de la Virgen
    date(2025, 10, 13),  # Día de la Raza
    date(2025, 11, 3),   # Todos los Santos
    date(2025, 11, 17),  # Independencia de Cartagena
    date(2025, 12, 8),   # Inmaculada Concepción
    date(2025, 12, 25),  # Navidad
    # 2026
    date(2026, 1, 1),    # Año Nuevo
    date(2026, 1, 12),   # Día de los Reyes Magos
    date(2026, 3, 23),   # Lunes Santo
    date(2026, 4, 2),    # Jueves Santo
    date(2026, 4, 3),    # Viernes Santo
    date(2026, 5, 1),    # Día del Trabajo
    date(2026, 6, 8),    # Ascensión del Señor
    date(2026, 6, 29),   # Corpus Christi
    date(2026, 7, 6),    # Sagrado Corazón
    date(2026, 7, 20),   # Día de la Independencia
    date(2026, 8, 7),    # Batalla de Boyacá
    date(2026, 8, 24),   # Asunción de la Virgen
    date(2026, 10, 12),  # Día de la Raza
    date(2026, 11, 2),   # Todos los Santos
    date(2026, 11, 16),  # Independencia de Cartagena
    date(2026, 12, 8),   # Inmaculada Concepción
    date(2026, 12, 25),  # Navidad
}


def es_dia_habil(fecha: date) -> bool:
    return fecha.weekday() < 5 and fecha not in COLOMBIAN_HOLIDAYS


def dias_habiles_entre(fecha_inicio: date, fecha_fin: date) -> int:
    count = 0
    actual = fecha_inicio
    while actual < fecha_fin:
        if es_dia_habil(actual):
            count += 1
        actual += timedelta(days=1)
    return count


def agregar_dias_habiles(fecha: date, n: int) -> date:
    resultado = fecha
    agregados = 0
    while agregados < n:
        resultado += timedelta(days=1)
        if es_dia_habil(resultado):
            agregados += 1
    return resultado

File: frontend/src/app/core/utils/business-days.ts
const COLOMBIAN_HOLIDAYS: Set<string> = new Set([
  // 2025
  '2025-01-01', '2025-01-06', '2025-03-24', '2025-04-17', '2025-04-18',
  '2025-05-01', '2025-05-26', '2025-06-16', '2025-06-23', '2025-06-29',
  '2025-07-20', '2025-08-07', '2025-08-18', '2025-10-13', '2025-11-03',
  '2025-11-17', '2025-12-08', '2025-12-25',
  // 2026
  '2026-01-01', '2026-01-12', '2026-03-23', '2026-04-02', '2026-04-03',
  '2026-05-01', '2026-06-08', '2026-06-29', '2026-07-06', '2026-07-20',
  '2026-08-07', '2026-08-24', '2026-10-12', '2026-11-02', '2026-11-16',
  '2026-12-08', '2026-12-25',
]);

function toKey(fecha: Date): string {
  const y = fecha.getFullYear();
  const m = String(fecha.getMonth() + 1).padStart(2, '0');
  const d = String(fecha.getDate()).padStart(2, '0');
  return `${y}-${m}-${d}`;
}

export function esDiaHabil(fecha: Date): boolean {
  return fecha.getDay() < 5 && !COLOMBIAN_HOLIDAYS.has(toKey(fecha));
}

export function diasHabilesEntre(inicio: Date, fin: Date): number {
  let count = 0;
  const actual = new Date(inicio);
  while (actual < fin) {
    if (esDiaHabil(actual)) count++;
    actual.setDate(actual.getDate() + 1);
  }
  return count;
}

export function agregarDiasHabiles(fecha: Date, n: number): Date {
  const resultado = new Date(fecha);
  let agregados = 0;
  while (agregados < n) {
    resultado.setDate(resultado.getDate() + 1);
    if (esDiaHabil(resultado)) agregados++;
  }
  return resultado;
}
